File: android/app/src/main/kotlin/com/edgelink/core/Envelope.kt
package com.edgelink.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

@Serializable
data class Envelope<T>(
    val t: String,
    val b: T
)

@Serializable
object EmptyBody

object EnvelopeTypes {
    const val STATUS_PING = "status.ping"
    const val STATUS_PONG = "status.pong"
    const val STATUS_CAPS = "status.caps"
    const val INPUT_POINTER = "input.pointer"
    const val INPUT_KEY = "input.key"
    const val INPUT_TEXT = "input.text"
    const val SCREEN_START = "screen.start"
    const val SCREEN_STOP = "screen.stop"
    const val SCREEN_META = "screen.meta"
    const val SCREEN_VIEWER_VISIBILITY = "screen.viewerVisibility"
    const val CTRL_POINTER = "ctrl.pointer"
    const val CTRL_KEY = "ctrl.key"
    const val CTRL_TEXT = "ctrl.text"
    const val CTRL_GLOBAL = "ctrl.global"
    const val RTC_OFFER = "rtc.offer"
    const val RTC_ANSWER = "rtc.answer"
    const val RTC_ICE = "rtc.ice"
    const val CLIPBOARD_SET = "clipboard.set"
    const val CLIPBOARD_HISTORY_REQUEST = "clipboard.history.request"
    const val CLIPBOARD_HISTORY_RESPONSE = "clipboard.history.response"
    const val NOTIFICATION_POST = "notification.post"
    const val NOTIFICATION_REMOVE = "notification.remove"
    const val SMS_MESSAGE = "sms.message"
    const val SMS_SEND = "sms.send"
    const val SMS_SEND_RESULT = "sms.send.result"
    const val PHONE_ACTION = "phone.action"
    const val PHONE_ACTION_RESULT = "phone.action.result"
    const val PHONE_RELAY_START = "phone.relay.start"
    const val PHONE_RELAY_ENDPOINT = "phone.relay.endpoint"
    const val PHONE_RELAY_MEDIA = "phone.relay.media"
    const val PHONE_CALL_STATUS = "phone.call.status"
    const val MILINK_STATUS = "milink.status"
    const val MILINK_FRAME = "milink.frame"
    const val MILINK_MIRROR_MEDIA = "milink.mirror.media"
    const val MILINK_COMMAND = "milink.command"
    const val MILINK_COMMAND_RESULT = "milink.command.result"
    const val ANDROID_MIC_STATUS = "android.mic.status"
    const val MAC_SLEEP = "mac.sleep"
    const val MAC_AWAKE = "mac.awake"
    const val TUNNEL_OPEN = "tunnel.open"
    const val TUNNEL_OPEN_RESULT = "tunnel.open.result"
    const val TUNNEL_DATA = "tunnel.data"
    const val TUNNEL_CLOSE = "tunnel.close"
    const val TUNNEL_ERROR = "tunnel.error"
    const val TUNNEL_FLOW = "tunnel.flow"
    const val BATTERY_STATUS = "battery.status"
}

object EnvelopeCodec {
    val json: Json = Json {
        encodeDefaults = true
        ignoreUnknownKeys = true
    }

    inline fun <reified T> encode(type: String, body: T): ByteArray =
        json.encodeToString(Envelope(type, body)).encodeToByteArray()

    inline fun <reified T> decode(bytes: ByteArray): Envelope<T> =
        json.decodeFromString(bytes.decodeToString())

    fun type(bytes: ByteArray): String =
        json.decodeFromString<Envelope<JsonObject>>(bytes.decodeToString()).t
}

@Serializable
data class InputPointerBody(
    val dx: Double = 0.0,
    val dy: Double = 0.0,
    val scrollX: Double? = null,
    val scrollY: Double? = null,
    val btn: String? = null
)

@Serializable
data class InputKeyBody(
    val key: String,
    val mods: List<String> = emptyList()
)

@Serializable
data class InputTextBody(
    val text: String
)

@Serializable
data class ScreenMetaBody(
    val w: Int,
    val h: Int,
    val scale: Double,
    val dpi: Int
)

@Serializable
data class ScreenViewerVisibilityBody(
    val visible: Boolean
)

@Serializable
data class CtrlPointerBody(
    val x: Int,
    val y: Int,
    val action: String,
    val wheelDy: Int? = null
)

@Serializable
data class CtrlKeyBody(
    val key: String,
    val down: Boolean,
    val mods: List<String> = emptyList()
)

@Serializable
data class CtrlTextBody(
    val text: String
)

@Serializable
data class CtrlGlobalBody(
    val action: String
)

@Serializable
data class RtcSdpBody(
    val sdp: String
)

@Serializable
data class RtcIceBody(
    val mid: String,
    val index: Int,
    val candidate: String
)

@Serializable
data class ClipboardSetBody(
    val text: String,
    val ts: Long,
    val hash: String,
    val kind: String? = null,
    val thumbnailBase64: String? = null,
    val sourceDeviceId: String? = null
)

enum class ClipboardKind(val intValue: Int) {
    TEXT(0),
    IMAGE(1),
    HTML(2),
    FILE(3);

    val wireName: String get() = name.lowercase()

    companion object {
        fun fromInt(value: Int): ClipboardKind? =
            entries.firstOrNull { it.intValue == value }

        fun fromWire(name: String?): ClipboardKind? =
            name?.let { needle -> entries.firstOrNull { it.wireName == needle } }
    }
}

@Serializable
data class StatusCapsBody(
    val clipboardHistory: Boolean = true,
    val clipboardThumbnail: Boolean = true
)

@Serializable
data class ClipboardHistoryRequestBody(
    val sinceTs: Long? = null,
    val limit: Int? = null
)

@Serializable
data class ClipboardHistoryItemBody(
    val id: String,
    val kind: String,
    val ts: Long,
    val hash: String,
    val text: String? = null,
    val thumbnailBase64: String? = null,
    val sourceDeviceId: String? = null
)

@Serializable
data class ClipboardHistoryResponseBody(
    val items: List<ClipboardHistoryItemBody> = emptyList()
)

@Serializable
data class NotificationPostBody(
    val id: String,
    val sourceDeviceId: String? = null,
    val sourcePlatform: String? = null,
    val app: String,
    val bundle: String? = null,
    val iconPngBase64: String? = null,
    val title: String,
    val text: String,
    val subtitle: String? = null,
    val ts: Long
)

@Serializable
data class NotificationRemoveBody(
    val id: String,
    val sourceDeviceId: String? = null
)

@Serializable
data class SmsMessageBody(
    val id: String,
    val sourceDeviceId: String? = null,
    val sourcePlatform: String? = null,
    val address: String,
    val text: String,
    val direction: String,
    val isBackfill: Boolean = false,
    val ts: Long
)

@Serializable
data class SmsSendBody(
    val requestId: String,
    val to: String,
    val text: String
)

@Serializable
data class SmsSendResultBody(
    val requestId: String,
    val to: String,
    val success: Boolean,
    val error: String? = null,
    val ts: Long
)

@Serializable
data class PhoneActionBody(
    val requestId: String,
    val action: String,
    val number: String? = null,
    val relayHost: String? = null,
    val relayPort: Int? = null,
    val relaySessionId: String? = null,
    val relayControlPort: Int? = null,
    val lanHost: String? = null,
    val lanPort: Int? = null,
    val lanProbePort: Int? = null
)

@Serializable
data class PhoneActionResultBody(
    val requestId: String,
    val action: String,
    val success: Boolean,
    val error: String? = null,
    val ts: Long
)

@Serializable
data class PhoneRelayStartRequestBody(
    val requestId: String,
    val reason: String,
    val ts: Long
)

@Serializable
data class PhoneRelayEndpointBody(
    val requestId: String,
    val relayHost: String? = null,
    val relayPort: Int? = null,
    val relaySessionId: String? = null,
    val relayControlPort: Int? = null,
    val lanHost: String? = null,
    val lanPort: Int? = null,
    val lanProbePort: Int? = null,
    val success: Boolean = true,
    val error: String? = null,
    val ts: Long
)

@Serializable
data class PhoneRelayMediaBody(
    val sessionId: String,
    val direction: String,
    val kind: String,
    val dataBase64: String? = null,
    val bytes: Int? = null,
    val sequence: Int? = null,
    val event: String? = null,
    val ts: Long
)

@Serializable
data class MiLinkMirrorMediaBody(
    val sessionId: String,
    val direction: String,
    val kind: String,
    val dataBase64: String? = null,
    val bytes: Int? = null,
    val sequence: Int? = null,
    val event: String? = null,
    val ts: Long
)

@Serializable
data class PhoneCallStatusBody(
    val callId: String,
    val state: String,
    val handle: String? = null,
    val displayName: String? = null,
    val direction: String? = null,
    val canAnswer: Boolean = false,
    val canHangUp: Boolean = false,
    val isActive: Boolean = false,
    val reason: String,
    val ts: Long
)

@Serializable
data class AndroidMicStatusBody(
    val active: Boolean,
    val source: Int? = null,
    val sourceName: String? = null,
    val sessionId: Int? = null,
    val silenced: Boolean? = null,
    val activeRecordingCount: Int = 0,
    val reason: String,
    val ts: Long
)

@Serializable
data class MiLinkServiceCapabilityBody(
    val id: String,
    val packageName: String,
    val appName: String,
    val serviceName: String,
    val category: String,
    val route: String,
    val available: Boolean,
    val preferred: Boolean = false,
    val bindAction: String? = null,
    val evidence: String
)

@Serializable
data class MiLinkStatusBody(
    val sourceDeviceId: String? = null,
    val sourcePlatform: String = "android",
    val route: String = "edgelink.secure",
    val officialDiscoveryRequired: Boolean = false,
    val available: Boolean,
    val rootProbeOk: Boolean,
    val attributionProbeOk: Boolean,
    val messengerTransportOk: Boolean,
    val castServiceOk: Boolean,
    val phoneContinuityOk: Boolean = false,
    val phoneCallRelayServiceOk: Boolean = false,
    val phoneMediaRelayCallbackOk: Boolean = false,
    val phoneRemoteDeviceCount: Int = 0,
    val phoneMediaRelayCandidateCount: Int = 0,
    val services: List<MiLinkServiceCapabilityBody> = emptyList(),
    val preferredRoutes: Map<String, String> = emptyMap(),
    val summary: String,
    val ts: Long
)

@Serializable
data class MiLinkFrameBody(
    val sourceDeviceId: String? = null,
    val sourcePlatform: String = "android",
    val route: String = "edgelink.secure",
    val clientNo: String,
    val sequence: Int,
    val dataBase64: String,
    val bytes: Int,
    val hasNext: Boolean,
    val ts: Long
)

@Serializable
data class MiLinkCommandBody(
    val requestId: String,
    val command: String,
    val args: Map<String, String> = emptyMap(),
    val ts: Long
)

@Serializable
data class MiLinkCommandResultBody(
    val requestId: String,
    val command: String,
    val success: Boolean,
    val route: String,
    val message: String,
    val data: Map<String, String> = emptyMap(),
    val ts: Long
)

@Serializable
data class BatteryStatusBody(
    val level: Int,
    val charging: Boolean,
    val plugged: String? = null,
    val temperature: Double? = null,
    val ts: Long
)

File: mac/Sources/EdgeLinkKit/Envelope.swift
import Foundation

public struct Envelope<Body: Codable & Sendable>: Codable, Sendable {
    public let t: String
    public let b: Body

    public init(t: String, b: Body) {
        self.t = t
        self.b = b
    }
}

public struct EmptyBody: Codable, Sendable {
    public init() {}
}

public enum EnvelopeType {
    public static let statusPing = "status.ping"
    public static let statusPong = "status.pong"
    public static let inputPointer = "input.pointer"
    public static let inputKey = "input.key"
    public static let inputText = "input.text"
    public static let screenStart = "screen.start"
    public static let screenStop = "screen.stop"
    public static let screenMeta = "screen.meta"
    public static let screenViewerVisibility = "screen.viewerVisibility"
    public static let ctrlPointer = "ctrl.pointer"
    public static let ctrlKey = "ctrl.key"
    public static let ctrlText = "ctrl.text"
    public static let ctrlGlobal = "ctrl.global"
    public static let rtcOffer = "rtc.offer"
    public static let rtcAnswer = "rtc.answer"
    public static let rtcIce = "rtc.ice"
    public static let clipboardSet = "clipboard.set"
    public static let notificationPost = "notification.post"
    public static let notificationRemove = "notification.remove"
    public static let smsMessage = "sms.message"
    public static let smsSend = "sms.send"
    public static let smsSendResult = "sms.send.result"
    public static let phoneAction = "phone.action"
    public static let phoneActionResult = "phone.action.result"
    public static let phoneRelayStart = "phone.relay.start"
    public static let phoneRelayEndpoint = "phone.relay.endpoint"
    public static let phoneCallStatus = "phone.call.status"
    public static let miLinkStatus = "milink.status"
    public static let miLinkFrame = "milink.frame"
    public static let miLinkCommand = "milink.command"
    public static let miLinkCommandResult = "milink.command.result"
    public static let androidMicStatus = "android.mic.status"
}

public struct InputPointerBody: Codable, Equatable, Sendable {
    public let dx: Double
    public let dy: Double
    public let scrollX: Double?
    public let scrollY: Double?
    public let btn: String?

    public init(dx: Double = 0, dy: Double = 0, scrollX: Double? = nil, scrollY: Double? = nil, btn: String? = nil) {
        self.dx = dx
        self.dy = dy
        self.scrollX = scrollX
        self.scrollY = scrollY
        self.btn = btn
    }
}

public struct InputKeyBody: Codable, Equatable, Sendable {
    public let key: String
    public let mods: [String]

    public init(key: String, mods: [String] = []) {
        self.key = key
        self.mods = mods
    }
}

public struct InputTextBody: Codable, Equatable, Sendable {
    public let text: String

    public init(text: String) {
        self.text = text
    }
}

public struct ScreenMetaBody: Codable, Equatable, Sendable {
    public let w: Int
    public let h: Int
    public let scale: Double
    public let dpi: Int

    public init(w: Int, h: Int, scale: Double, dpi: Int) {
        self.w = w
        self.h = h
        self.scale = scale
        self.dpi = dpi
    }
}

public struct ScreenViewerVisibilityBody: Codable, Equatable, Sendable {
    public let visible: Bool

    public init(visible: Bool) {
        self.visible = visible
    }
}

public struct CtrlPointerBody: Codable, Equatable, Sendable {
    public let x: Int
    public let y: Int
    public let action: String
    public let wheelDy: Int?

    public init(x: Int, y: Int, action: String, wheelDy: Int? = nil) {
        self.x = x
        self.y = y
        self.action = action
        self.wheelDy = wheelDy
    }
}

public struct CtrlKeyBody: Codable, Equatable, Sendable {
    public let key: String
    public let down: Bool
    public let mods: [String]

    public init(key: String, down: Bool, mods: [String] = []) {
        self.key = key
        self.down = down
        self.mods = mods
    }
}

public struct CtrlTextBody: Codable, Equatable, Sendable {
    public let text: String

    public init(text: String) {
        self.text = text
    }
}

public struct CtrlGlobalBody: Codable, Equatable, Sendable {
    public let action: String

    public init(action: String) {
        self.action = action
    }
}

public struct RtcSdpBody: Codable, Equatable, Sendable {
    public let sdp: String

    public init(sdp: String) {
        self.sdp = sdp
    }
}

public struct RtcIceBody: Codable, Equatable, Sendable {
    public let mid: String
    public let index: Int
    public let candidate: String

    public init(mid: String, index: Int, candidate: String) {
        self.mid = mid
        self.index = index
        self.candidate = candidate
    }
}

public struct ClipboardSetBody: Codable, Equatable, Sendable {
    public let text: String
    public let ts: Int64
    public let hash: String

    public init(text: String, ts: Int64, hash: String) {
        self.text = text
        self.ts = ts
        self.hash = hash
    }
}

public struct NotificationPostBody: Codable, Equatable, Sendable {
    public let id: String
    public let sourceDeviceId: String?
    public let sourcePlatform: String?
    public let app: String
    public let bundle: String?
    public let iconPngBase64: String?
    public let title: String
    public let text: String
    public let subtitle: String?
    public let ts: Int64

    public init(
        id: String,
        sourceDeviceId: String? = nil,
        sourcePlatform: String? = nil,
        app: String,
        bundle: String? = nil,
        iconPngBase64: String? = nil,
        title: String,
        text: String,
        subtitle: String? = nil,
        ts: Int64
    ) {
        self.id = id
        self.sourceDeviceId = sourceDeviceId
        self.sourcePlatform = sourcePlatform
        self.app = app
        self.bundle = bundle
        self.iconPngBase64 = iconPngBase64
        self.title = title
        self.text = text
        self.subtitle = subtitle
        self.ts = ts
    }
}

public struct NotificationRemoveBody: Codable, Equatable, Sendable {
    public let id: String
    public let sourceDeviceId: String?

    public init(id: String, sourceDeviceId: String? = nil) {
        self.id = id
        self.sourceDeviceId = sourceDeviceId
    }
}

public struct SmsMessageBody: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let sourceDeviceId: String?
    public let sourcePlatform: String?
    public let address: String
    public let text: String
    public let direction: String
    public let isBackfill: Bool
    public let ts: Int64

    public init(
        id: String,
        sourceDeviceId: String? = nil,
        sourcePlatform: String? = nil,
        address: String,
        text: String,
        direction: String,
        isBackfill: Bool = false,
        ts: Int64
    ) {
        self.id = id
        self.sourceDeviceId = sourceDeviceId
        self.sourcePlatform = sourcePlatform
        self.address = address
        self.text = text
        self.direction = direction
        self.isBackfill = isBackfill
        self.ts = ts
    }
}

public struct SmsSendBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let to: String
    public let text: String

    public init(requestId: String, to: String, text: String) {
        self.requestId = requestId
        self.to = to
        self.text = text
    }
}

public struct SmsSendResultBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let to: String
    public let success: Bool
    public let error: String?
    public let ts: Int64

    public init(requestId: String, to: String, success: Bool, error: String? = nil, ts: Int64) {
        self.requestId = requestId
        self.to = to
        self.success = success
        self.error = error
        self.ts = ts
    }
}

public struct PhoneActionBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let action: String
    public let number: String?
    public let relayHost: String?
    public let relayPort: Int?
    public let relaySessionId: String?
    public let relayControlPort: Int?

    public init(
        requestId: String,
        action: String,
        number: String? = nil,
        relayHost: String? = nil,
        relayPort: Int? = nil,
        relaySessionId: String? = nil,
        relayControlPort: Int? = nil
    ) {
        self.requestId = requestId
        self.action = action
        self.number = number
        self.relayHost = relayHost
        self.relayPort = relayPort
        self.relaySessionId = relaySessionId
        self.relayControlPort = relayControlPort
    }
}

public struct PhoneActionResultBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let action: String
    public let success: Bool
    public let error: String?
    public let ts: Int64

    public init(requestId: String, action: String, success: Bool, error: String? = nil, ts: Int64) {
        self.requestId = requestId
        self.action = action
        self.success = success
        self.error = error
        self.ts = ts
    }
}

public struct PhoneRelayStartRequestBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let reason: String
    public let ts: Int64

    public init(requestId: String, reason: String, ts: Int64) {
        self.requestId = requestId
        self.reason = reason
        self.ts = ts
    }
}

public struct PhoneRelayEndpointBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let relayHost: String?
    public let relayPort: Int?
    public let relaySessionId: String?
    public let relayControlPort: Int?
    public let success: Bool
    public let error: String?
    public let ts: Int64

    public init(
        requestId: String,
        relayHost: String? = nil,
        relayPort: Int? = nil,
        relaySessionId: String? = nil,
        relayControlPort: Int? = nil,
        success: Bool = true,
        error: String? = nil,
        ts: Int64
    ) {
        self.requestId = requestId
        self.relayHost = relayHost
        self.relayPort = relayPort
        self.relaySessionId = relaySessionId
        self.relayControlPort = relayControlPort
        self.success = success
        self.error = error
        self.ts = ts
    }
}

public struct PhoneCallStatusBody: Codable, Equatable, Sendable {
    public let callId: String
    public let state: String
    public let handle: String?
    public let displayName: String?
    public let direction: String?
    public let canAnswer: Bool
    public let canHangUp: Bool
    public let isActive: Bool
    public let reason: String
    public let ts: Int64

    public init(
        callId: String,
        state: String,
        handle: String? = nil,
        displayName: String? = nil,
        direction: String? = nil,
        canAnswer: Bool = false,
        canHangUp: Bool = false,
        isActive: Bool = false,
        reason: String,
        ts: Int64
    ) {
        self.callId = callId
        self.state = state
        self.handle = handle
        self.displayName = displayName
        self.direction = direction
        self.canAnswer = canAnswer
        self.canHangUp = canHangUp
        self.isActive = isActive
        self.reason = reason
        self.ts = ts
    }
}

public struct AndroidMicStatusBody: Codable, Equatable, Sendable {
    public let active: Bool
    public let source: Int?
    public let sourceName: String?
    public let sessionId: Int?
    public let silenced: Bool?
    public let activeRecordingCount: Int
    public let reason: String
    public let ts: Int64

    public init(
        active: Bool,
        source: Int? = nil,
        sourceName: String? = nil,
        sessionId: Int? = nil,
        silenced: Bool? = nil,
        activeRecordingCount: Int = 0,
        reason: String,
        ts: Int64
    ) {
        self.active = active
        self.source = source
        self.sourceName = sourceName
        self.sessionId = sessionId
        self.silenced = silenced
        self.activeRecordingCount = activeRecordingCount
        self.reason = reason
        self.ts = ts
    }
}

public struct MiLinkServiceCapabilityBody: Codable, Equatable, Sendable {
    public let id: String
    public let packageName: String
    public let appName: String
    public let serviceName: String
    public let category: String
    public let route: String
    public let available: Bool
    public let preferred: Bool
    public let bindAction: String?
    public let evidence: String

    public init(
        id: String,
        packageName: String,
        appName: String,
        serviceName: String,
        category: String,
        route: String,
        available: Bool,
        preferred: Bool = false,
        bindAction: String? = nil,
        evidence: String
    ) {
        self.id = id
        self.packageName = packageName
        self.appName = appName
        self.serviceName = serviceName
        self.category = category
        self.route = route
        self.available = available
        self.preferred = preferred
        self.bindAction = bindAction
        self.evidence = evidence
    }
}

public struct MiLinkStatusBody: Codable, Equatable, Sendable {
    public let sourceDeviceId: String?
    public let sourcePlatform: String
    public let route: String
    public let officialDiscoveryRequired: Bool
    public let available: Bool
    public let rootProbeOk: Bool
    public let attributionProbeOk: Bool
    public let messengerTransportOk: Bool
    public let castServiceOk: Bool
    public let phoneContinuityOk: Bool?
    public let phoneCallRelayServiceOk: Bool?
    public let phoneMediaRelayCallbackOk: Bool?
    public let phoneRemoteDeviceCount: Int?
    public let phoneMediaRelayCandidateCount: Int?
    public let services: [MiLinkServiceCapabilityBody]?
    public let preferredRoutes: [String: String]?
    public let summary: String
    public let ts: Int64

    public init(
        sourceDeviceId: String? = nil,
        sourcePlatform: String = "android",
        route: String = "edgelink.secure",
        officialDiscoveryRequired: Bool = false,
        available: Bool,
        rootProbeOk: Bool,
        attributionProbeOk: Bool,
        messengerTransportOk: Bool,
        castServiceOk: Bool,
        phoneContinuityOk: Bool? = nil,
        phoneCallRelayServiceOk: Bool? = nil,
        phoneMediaRelayCallbackOk: Bool? = nil,
        phoneRemoteDeviceCount: Int? = nil,
        phoneMediaRelayCandidateCount: Int? = nil,
        services: [MiLinkServiceCapabilityBody]? = nil,
        preferredRoutes: [String: String]? = nil,
        summary: String,
        ts: Int64
    ) {
        self.sourceDeviceId = sourceDeviceId
        self.sourcePlatform = sourcePlatform
        self.route = route
        self.officialDiscoveryRequired = officialDiscoveryRequired
        self.available = available
        self.rootProbeOk = rootProbeOk
        self.attributionProbeOk = attributionProbeOk
        self.messengerTransportOk = messengerTransportOk
        self.castServiceOk = castServiceOk
        self.phoneContinuityOk = phoneContinuityOk
        self.phoneCallRelayServiceOk = phoneCallRelayServiceOk
        self.phoneMediaRelayCallbackOk = phoneMediaRelayCallbackOk
        self.phoneRemoteDeviceCount = phoneRemoteDeviceCount
        self.phoneMediaRelayCandidateCount = phoneMediaRelayCandidateCount
        self.services = services
        self.preferredRoutes = preferredRoutes
        self.summary = summary
        self.ts = ts
    }
}

public struct MiLinkFrameBody: Codable, Equatable, Sendable {
    public let sourceDeviceId: String?
    public let sourcePlatform: String
    public let route: String
    public let clientNo: String
    public let sequence: Int
    public let dataBase64: String
    public let bytes: Int
    public let hasNext: Bool
    public let ts: Int64

    public init(
        sourceDeviceId: String? = nil,
        sourcePlatform: String = "android",
        route: String = "edgelink.secure",
        clientNo: String,
        sequence: Int,
        dataBase64: String,
        bytes: Int,
        hasNext: Bool,
        ts: Int64
    ) {
        self.sourceDeviceId = sourceDeviceId
        self.sourcePlatform = sourcePlatform
        self.route = route
        self.clientNo = clientNo
        self.sequence = sequence
        self.dataBase64 = dataBase64
        self.bytes = bytes
        self.hasNext = hasNext
        self.ts = ts
    }
}

public struct MiLinkCommandBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let command: String
    public let args: [String: String]
    public let ts: Int64

    public init(
        requestId: String,
        command: String,
        args: [String: String] = [:],
        ts: Int64
    ) {
        self.requestId = requestId
        self.command = command
        self.args = args
        self.ts = ts
    }
}

public struct MiLinkCommandResultBody: Codable, Equatable, Sendable {
    public let requestId: String
    public let command: String
    public let success: Bool
    public let route: String
    public let message: String
    public let data: [String: String]
    public let ts: Int64

    public init(
        requestId: String,
        command: String,
        success: Bool,
        route: String,
        message: String,
        data: [String: String] = [:],
        ts: Int64
    ) {
        self.requestId = requestId
        self.command = command
        self.success = success
        self.route = route
        self.message = message
        self.data = data
        self.ts = ts
    }
}
